crates/bitwarden/src/platform/fido2/authenticator.rs

```diff
@@ -92,242 +90,373 @@ impl<'a> Fido2Authenticator<'a> {
+                    .transpose()?,
+                extensions: request
+                    .extensions
+                    .map(|e| serde_json::from_str(&e))
+                    .transpose()?,
+                options: passkey::types::ctap2::make_credential::Options {
+                    rk: request.options.rk,
+                    up: true,
+                    uv: request.options.uv != UV::Discouraged,
                 },
-            },
+                pin_auth: None,
+                pin_protocol: None,
+            })
+            .await;
+
+        let response = match response {
+            Ok(x) => x,
+            Err(e) => return Err(format!("get_assertion error: {e:?}").into()),
+        };
+
+        let authenticator_data = response.auth_data.to_vec();
+        let credential_id = response
+            .auth_data
+            .attested_credential_data
+            .ok_or("Missing attested_credential_data")?
+            .credential_id()
+            .to_vec();
+
+        Ok(GetAssertionResult {
+            credential_id,
+            authenticator_data,
+            signature: response.signature.into(),
+            user_handle: response.user.ok_or("Missing user")?.id.into(),
+            selected_credential: self.get_selected_credential()?,
         })
     }
 
-    // TODO: Fido2CredentialView contains all the fields, maybe we need a Fido2CredentialListView?
+    // TODO(Fido2): We need some changes in the passkey crate to do this correctly
     pub async fn silently_discover_credentials(
         &mut self,
         rp_id: String,
     ) -> Result<Vec<Fido2CredentialView>> {
-        Ok(vec![])
+        let result = self.credential_store.find_credentials(None, rp_id).await?;
+        Ok(result
+            .into_iter()
+            .filter_map(|c| c.login?.fido2_credentials)
+            .flatten()
+            .collect())
     }
 
-    pub(crate) fn to_user_interface(&'a self) -> UserInterfaceImpl<'_> {
-        UserInterfaceImpl {
-            authenticator: self,
-        }
+    pub(super) fn get_authenticator(
+        &self,
+    ) -> Authenticator<CredentialStoreImpl, UserValidationMethodImpl> {
+        Authenticator::new(
+            AAGUID,
+            CredentialStoreImpl {
+                authenticator: self,
+            },
+            UserValidationMethodImpl {
+                authenticator: self,
+            },
+        )
     }
-    pub(crate) fn to_credential_store(&'a self) -> CredentialStoreImpl<'_> {
-        CredentialStoreImpl {
-            authenticator: self,
-        }
+
+    pub(super) fn get_selected_credential(&self) -> Result<SelectedCredential> {
+        let cipher = self
+            .selected_credential
+            .lock()
+            .expect("Mutex is not poisoned")
+            .take()
+            .ok_or("No selected credential available")?;
+
+        let login = require!(cipher.login.as_ref());
+        let creds = require!(login.fido2_credentials.as_ref());
```

Comment:
@coroiu this is a custom macro, you can find the definition [here](https://github.com/bitwarden/sdk/blob/3065ba8f084039089109862c8c04f2855fcf6a8e/crates/bitwarden/src/error.rs#L154) 😄 There is a good doc comment for this macro:

```
/// This macro is used to require that a value is present or return an error otherwise.
/// It is equivalent to using `val.ok_or(Error::MissingFields)?`, but easier to use and
/// with a more descriptive error message.
/// Note that this macro will return early from the function if the value is not present.
```

Reply:
Yep that's right, it expands to this:
```rust
        let login = match cipher.login.as_ref() {
            Some(val) => val,
            None => return Err(Error::MissingFields("cipher.login")),
        };
```

The macro mostly exists because the `-api-*` generated crates have a lot of `Option<>` values that we want to handle.

In this case, those fields should be always defined there, so we could have just used `.expect()` as well.

Reply:
Ah of course, no wonder I couldn't find it by googling for it 🤦 Thank you!

> In this case, those fields should be always defined there, so we could have just used .expect() as well.

The difference would just be that `expect()` panics while `require!` returns an `Err`. Would it make sense to switch to `expect()`? Since any error here would be unexpected and there's no real way of properly handling it? Or is that not the comment way to think about the difference between panic and err ?

Reply:
`panic` kills the whole app. If there is no way to recover from an error `panic` is appropriate, but is essentially the same as crashing the app